File: debop4s-data/src/main/scala/com/github/debop4s/data/model/PersistentObject.scala
package com.github.debop4s.data.model

import com.github.debop4s.core.ValueObject
import com.github.debop4s.core.utils.ToStringHelper
import javax.persistence.Transient

/**
 * com.github.debop4s.data.model.PersistentObject
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 12. 24. 오후 2:42
 */
@SerialVersionUID(-7066710546641101707L)
trait PersistentObject extends ValueObject {

  @Transient
  private var persisted: Boolean = false

  def isPersisted = persisted

  protected def setPersisted(v: Boolean) {
    persisted = v
  }

  def onPersist() {
    persisted = true
  }

  def onLoad() {
    persisted = true
  }

  override protected def buildStringHelper: ToStringHelper =
    super.buildStringHelper
    .add("persisted", persisted)
}

File: debop4s-data-orm/src/main/java/debop4s/data/orm/model/PersistentObject.java
package debop4s.data.orm.model;

import debop4s.core.ValueObject;

/**
 * Hibernate, JPA 용 데이터를 표현하는 객체의 기본 클래스 (Entity나 Component의 조상입니다)
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 6. 30. 오후 12:30
 */
public interface PersistentObject extends ValueObject {

    /**
     * 엔티티가 영구 저장되었는지 여부
     *
     * @return true 면 영구 저장, false면 transient object 입니다.
     */
    boolean isPersisted();

    /** 엔티티 저장 시 호출되는 메소드 */
    void onSave();

    /** 엔티티 저장 시 호출되는 메소드 */
    void onPersist();

    /** 엔티티를 영구 저장소에서 로드 시에 호출되는 메소드 */
    void onLoad();
}
